File: research/tt/strategies/time_series_momentum.py
"""``time_series_momentum``, exactly as docs/CONTRACTS.md defines it.

Time-series momentum (Moskowitz, Ooi and Pedersen, *Time Series Momentum*,
2012; Hurst, Ooi and Pedersen, *A Century of Evidence on Trend-Following
Investing*, 2017), long-only, one sleeve per risk asset: every
``rebalance_days`` bars each asset is held if its own trailing
``lookback``-bar return is positive and rests in the haven — the universe's
last symbol — otherwise. Unlike ``dual_momentum`` no sleeve looks at another
asset or at the haven's return; the threshold is zero and is not a parameter.

The Go runtime implements the same text; ``golden/<name>/expected_signals.csv``
holds the two to 1e-9. Change this and the contract and the goldens together.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd

from tt.indicators import rolling_return
from tt.strategies.ratio import align, split, weights_frame

NAME = "time_series_momentum"

KNOWN = {"lookback", "rebalance_days"}


@dataclass(frozen=True)
class Params:
    """The spec's params block, typed."""

    lookback: int
    rebalance_days: int

    @classmethod
    def from_dict(cls, d: dict[str, float]) -> Params:
        """Build from a spec's ``params`` mapping, refusing unknown keys."""
        unknown = set(d) - KNOWN
        if unknown:
            raise ValueError(f"unknown params {sorted(unknown)}")
        missing = KNOWN - set(d)
        if missing:
            raise ValueError(f"missing params {sorted(missing)}")
        p = cls(lookback=int(d["lookback"]), rebalance_days=int(d["rebalance_days"]))
        if p.lookback < 1 or p.lookback != d["lookback"]:
            raise ValueError("lookback must be a whole number >= 1")
        if p.rebalance_days < 1 or p.rebalance_days != d["rebalance_days"]:
            raise ValueError("rebalance_days must be a whole number >= 1")
        return p


def replay(
    bars: dict[str, pd.DataFrame], universe: list[str], params: Params, gross_leverage: float
) -> pd.DataFrame:
    """Replay the rebalance schedule over the aligned bars.

    Returns one row per aligned date with ``r_<symbol>`` (the trailing
    return, NaN while undefined) and ``s_<symbol>`` (1 held, 0 in the haven)
    for each risk asset, ``rebalance`` (1 on a rebalance bar), and
    ``w_<symbol>`` for every symbol.
    """
    risk, haven = split(universe)
    df = align(bars, universe)
    n = len(df)
    k = len(risk)
    lb, rb = params.lookback, params.rebalance_days
    out = pd.DataFrame({"date": df["date"]})
    rebalance = np.array([1 if t >= lb and (t - lb) % rb == 0 else 0 for t in range(n)], dtype=int)
    out["rebalance"] = rebalance
    in_risk = np.zeros(n, dtype=int)
    for s in risk:
        r = rolling_return(df[s].to_numpy(dtype=float), lb)
        st_col = np.zeros(n, dtype=int)
        st = 0
        for t in range(n):
            if rebalance[t]:
                st = 1 if r[t] > 0 else 0
            st_col[t] = st
        out[f"r_{s}"] = r
        out[f"s_{s}"] = st_col
        out[f"w_{s}"] = np.where(st_col == 1, gross_leverage / k, 0.0)
        in_risk += st_col
    flat = (k - in_risk).astype(float)
    out[f"w_{haven}"] = gross_leverage * flat / k
    return out


def signals(
    bars: dict[str, pd.DataFrame], universe: list[str], params: Params, gross_leverage: float
) -> pd.DataFrame:
    """Target weights in long form: ``date, symbol, target_weight``, every date."""
    tr = replay(bars, universe, params, gross_leverage)
    parts = [
        pd.DataFrame({"date": tr["date"], "symbol": s, "target_weight": tr[f"w_{s}"]})
        for s in universe
    ]
    return pd.concat(parts).sort_values(["date", "symbol"]).reset_index(drop=True)


def switches(replayed: pd.DataFrame, universe: list[str]) -> int:
    """How many times any sleeve changed side."""
    risk, _ = split(universe)
    return sum(int(np.count_nonzero(np.diff(replayed[f"s_{s}"].to_numpy()))) for s in risk)


__all__ = ["NAME", "Params", "align", "replay", "signals", "split", "switches", "weights_frame"]

File: server/internal/strategy/tsmomentum/time_series_momentum.go
// Package tsmomentum is the time_series_momentum strategy: every
// rebalance_days bars each risk asset is held if its own trailing
// lookback-bar return is positive and rests in the haven (the universe's
// last symbol) otherwise. No sleeve looks at another asset or at the haven's
// return, which is what separates it from dual_momentum; the threshold is
// zero and not a parameter.
//
// The definition is docs/CONTRACTS.md, implemented to the letter, because
// the same text is implemented in Python for research and a golden file
// checks the two against each other row by row (docs/PLAN.md §4.3).
package tsmomentum

import (
	"fmt"
	"math"
	"sort"
	"strings"
	"time"

	"github.com/chinmay28/bulls-and-bears/server/internal/bars"
	"github.com/chinmay28/bulls-and-bears/server/internal/indicator"
	"github.com/chinmay28/bulls-and-bears/server/internal/strategy"
	"github.com/chinmay28/bulls-and-bears/server/internal/strategy/sleeves"
)

// Name is what a spec's `strategy` field says to get this implementation.
const Name = "time_series_momentum"

func init() { strategy.Register(Name, New) }

// Params is the spec's params block, typed.
type Params struct {
	Lookback      int
	RebalanceDays int
}

// TSMomentum is the strategy for one universe. Immutable once built; the
// sleeve states are replayed from history on every call, never stored.
type TSMomentum struct {
	risk   []string
	haven  string
	params Params
	gross  float64
}

// Signal is one aligned bar: per risk asset the trailing return (Defined
// false while the window is short) and the sleeve's state after this bar,
// whether the bar was a rebalance bar, and the target weights.
type Signal struct {
	Date      time.Time
	Return    map[string]float64
	Defined   map[string]bool
	State     map[string]int
	Rebalance bool
	Weights   map[string]float64
}

// New is the strategy.Factory.
func New(universe []string, params strategy.Params, sizing strategy.Sizing) (strategy.Strategy, error) {
	return Parse(universe, params, sizing)
}

// Parse validates and builds a *TSMomentum. Every check is a spec error,
// not a default.
func Parse(universe []string, params strategy.Params, sizing strategy.Sizing) (*TSMomentum, error) {
	risk, haven, err := sleeves.Universe("tsmomentum", universe)
	if err != nil {
		return nil, err
	}
	if err := sleeves.Known("tsmomentum", params, "lookback", "rebalance_days"); err != nil {
		return nil, err
	}
	lb, err := sleeves.WholeNumber("tsmomentum", "lookback", params["lookback"], 1)
	if err != nil {
		return nil, err
	}
	rb, err := sleeves.WholeNumber("tsmomentum", "rebalance_days", params["rebalance_days"], 1)
	if err != nil {
		return nil, err
	}
	if !(sizing.GrossLeverage > 0) || math.IsInf(sizing.GrossLeverage, 0) {
		return nil, fmt.Errorf("tsmomentum: gross_leverage must be > 0, got %v", sizing.GrossLeverage)
	}
	return &TSMomentum{risk: risk, haven: haven, params: Params{Lookback: lb, RebalanceDays: rb}, gross: sizing.GrossLeverage}, nil
}

// Universe is the spec's order: the risk assets, then the haven.
func (m *TSMomentum) Universe() []string { return append(append([]string(nil), m.risk...), m.haven) }

// Params is what the strategy was built with.
func (m *TSMomentum) Params() Params { return m.params }

// Targets returns the weights for the last aligned bar. pos is ignored on
// purpose: the state is replayed from history alone.
func (m *TSMomentum) Targets(hist map[string][]bars.Bar, _ []strategy.Position) (map[string]float64, error) {
	sig, err := m.Signals(hist)
	if err != nil {
		return nil, err
	}
	if len(sig) == 0 {
		return nil, fmt.Errorf("tsmomentum: %v share no dates", m.Universe())
	}
	return sig[len(sig)-1].Weights, nil
}

// Signals replays the strategy over the aligned history, one Signal per
// aligned bar, oldest first.
func (m *TSMomentum) Signals(hist map[string][]bars.Bar) ([]Signal, error) {
	dates, px, err := sleeves.Align("tsmomentum", m.Universe(), m.haven, hist)
	if err != nil {
		return nil, err
	}
	n := len(dates)
	k := len(m.risk)
	out := make([]Signal, n)
	lb, rb := m.params.Lookback, m.params.RebalanceDays
	for t := range out {
		out[t] = Signal{Date: dates[t], Return: map[string]float64{}, Defined: map[string]bool{},
			State: map[string]int{}, Weights: map[string]float64{}, Rebalance: t >= lb && (t-lb)%rb == 0}
	}
	inRisk := make([]int, n)
	for _, sym := range m.risk {
		ret := indicator.RollingReturn(px[sym], lb)
		s := 0
		for t := 0; t < n; t++ {
			if out[t].Rebalance {
				s = 0
				if ret.Defined[t] && ret.Value[t] > 0 {
					s = 1
				}
			}
			out[t].Return[sym], out[t].Defined[sym], out[t].State[sym] = ret.Value[t], ret.Defined[t], s
			if s == 1 {
				out[t].Weights[sym] = m.gross / float64(k)
				inRisk[t]++
			} else {
				out[t].Weights[sym] = 0
			}
		}
	}
	for t := range out {
		out[t].Weights[m.haven] = m.gross * float64(k-inRisk[t]) / float64(k)
	}
	return out, nil
}

// Weights is the backtest engine's view of Signals.
func (m *TSMomentum) Weights(hist map[string][]bars.Bar) ([]map[string]float64, error) {
	sig, err := m.Signals(hist)
	if err != nil {
		return nil, err
	}
	out := make([]map[string]float64, len(sig))
	for i, s := range sig {
		out[i] = s.Weights
	}
	return out, nil
}

// String names the strategy and its parameters, for logs.
func (m *TSMomentum) String() string {
	return fmt.Sprintf("%s(%s | %s; lookback %d, every %d)", Name, strings.Join(m.risk, ","), m.haven, m.params.Lookback, m.params.RebalanceDays)
}

var _ = sort.Strings
